### ISSUE TITLE
`empty` and `non-empty` types

### ISSUE THREAD

**Issue by Ocramius** [azjezz/psl#222]
Working with some code today, I noticed that @michaelpetri implemented following in a code snippet, while trying to migrate from `webmozart/assert` to `azjezz/psl`:

```php
        $roles = Type\vec(
            Type\object(UserRole::class)
        )->coerce($roles);

        Assert::notEmpty($roles);
```

In practice, this is mostly a nuisance, but I feel like we could improve the `Psl\Type` namespace by adding `empty` and `!empty` type mutators, possibly usable via `Psl\Type\intersection()`:

```php
        $roles = Type\intersect(
            Type\vec(
                Type\object(UserRole::class)
            ),
            Type\not(Type\empty())
        )->coerce($roles);
```

Obviously, `Type\not()` is perhaps a bit of a stretch, but the general approach is to allow inferring `non-empty-list`, `non-empty-array`, `!0` (non-empty + integer), etc. from such a type.

**Comment by Ocramius** [comment]
Related: https://github.com/azjezz/psl/issues/145

**Comment by azjezz** [comment]
Hm, i think this is a good idea 👍

**Comment by azjezz** [comment]
I'm marking this as resolved since these types were implemented in #201, #200

**Comment by Ocramius** [comment]
🥳
